lib/tapioca/compilers/dsl/url_helpers.rb

```diff
@@ -0,0 +1,82 @@
+              mod.create_include("ActionDispatch::Routing::UrlFor")
+              mod.create_include("ActionDispatch::Routing::PolymorphicRoutes")
+
+              path_helper_methods.each do |method|
+                mod.create_method(
+                  method.to_s,
+                  parameters: [Parlour::RbiGenerator::Parameter.new("*args", type: "T.untyped")],
+                  return_type: "String"
+                )
+              end
+            end
+          elsif constant == GeneratedUrlHelpersModule
+            root.create_module("GeneratedUrlHelpersModule") do |mod|
+              mod.create_include("ActionDispatch::Routing::UrlFor")
+              mod.create_include("ActionDispatch::Routing::PolymorphicRoutes")
+
+              url_helper_methods.each do |method|
+                mod.create_method(
+                  method.to_s,
+                  parameters: [Parlour::RbiGenerator::Parameter.new("*args", type: "T.untyped")],
+                  return_type: "String"
+                )
+              end
+            end
+          else
+            root.path(constant) do |mod|
+              mod.create_include("GeneratedUrlHelpersModule") if
+                constant.ancestors.include?(GeneratedUrlHelpersModule)
+              mod.create_include("GeneratedPathHelpersModule") if
+                constant.ancestors.include?(GeneratedPathHelpersModule)
+              mod.create_extend("GeneratedUrlHelpersModule") if
+                constant.singleton_class.ancestors.include?(GeneratedUrlHelpersModule)
+              mod.create_extend("GeneratedPathHelpersModule") if
+                constant.singleton_class.ancestors.include?(GeneratedPathHelpersModule)
+            end
+          end
+        end
+
+        sig { override.returns(T::Enumerable[T.untyped]) }
+        def gather_constants
+          Object.const_set(:GeneratedUrlHelpersModule, Rails.application.routes.named_routes.url_helpers_module)
+          Object.const_set(:GeneratedPathHelpersModule, Rails.application.routes.named_routes.path_helpers_module)
+          ObjectSpace.each_object(Module).select do |mod|
+            (T.unsafe(mod).ancestors.include?(GeneratedUrlHelpersModule) &&
+             !T.unsafe(mod).try(:superclass)&.ancestors&.include?(GeneratedUrlHelpersModule)) ||
+            (T.unsafe(mod).singleton_class.ancestors.include?(GeneratedUrlHelpersModule) &&
+             !T.unsafe(mod).singleton_class.try(:superclass)&.ancestors&.include?(GeneratedPathHelpersModule)) ||
+            (T.unsafe(mod).ancestors.include?(GeneratedPathHelpersModule) &&
+             !T.unsafe(mod).try(:superclass)&.ancestors&.include?(GeneratedPathHelpersModule)) ||
+            (T.unsafe(mod).singleton_class.ancestors.include?(GeneratedPathHelpersModule) &&
+             !T.unsafe(mod).singleton_class.try(:superclass)&.ancestors&.include?(GeneratedPathHelpersModule))
+          end.select(&T.unsafe(:name)) << "ActionDispatch::IntegrationTest"
+        end
```

Comment:
```suggestion
          constants = ObjectSpace.each_object(Module).select do |mod|
            mod = T.let(mod, T.class_of(Module))
            includes_helper?(mod, GeneratedUrlHelpersModule)
     		  || includes_helper?(mod, GeneratedPathHelpersModule)
     		  || includes_helper?(mod.singleton_class, GeneratedUrlHelpersModule)
     		  || includes_helper?(mod.singleton_class, GeneratedPathHelpersModule)
          end.select(&:name)
          constants << "ActionDispatch::IntegrationTest"
          constants
        end
```
would be much nicer and readable.

```ruby
def includes_helper?(mod, helper)
  superclass_ancestors = [] 
  superclass_ancestors = mod.superclass.ancestors if Class === mod
  (mod.ancestors - superclass_ancestors).include?(helper)
end
```